File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/subscription/NoOpSubscriptionManager.java
package com.apollographql.apollo.internal.subscription;

import com.apollographql.apollo.api.Subscription;
import com.apollographql.apollo.subscription.OnSubscriptionManagerStateChangeListener;
import com.apollographql.apollo.subscription.SubscriptionManagerState;
import org.jetbrains.annotations.NotNull;

public final class NoOpSubscriptionManager implements SubscriptionManager {

  @Override
  public <T> void subscribe(@NotNull Subscription<?, T, ?> subscription, @NotNull Callback<T> callback) {
    throw new IllegalStateException("Subscription manager is not configured");
  }

  @Override
  public void unsubscribe(@NotNull Subscription<?, ?, ?> subscription) {
    throw new IllegalStateException("Subscription manager is not configured");
  }

  @Override
  public void start() {
    throw new IllegalStateException("Subscription manager is not configured");
  }

  @Override
  public void stop() {
    throw new IllegalStateException("Subscription manager is not configured");
  }

  @Override
  public SubscriptionManagerState getState() {
    return SubscriptionManagerState.DISCONNECTED;
  }

  @Override
  public void addOnStateChangeListener(@NotNull OnSubscriptionManagerStateChangeListener onStateChangeListener) {
    throw new IllegalStateException("Subscription manager is not configured");
  }

  @Override
  public void removeOnStateChangeListener(@NotNull OnSubscriptionManagerStateChangeListener onStateChangeListener) {
    throw new IllegalStateException("Subscription manager is not configured");
  }
}

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/subscription/NoOpSubscriptionManager.kt
package com.apollographql.apollo.internal.subscription

import com.apollographql.apollo.api.Subscription
import com.apollographql.apollo.subscription.OnSubscriptionManagerStateChangeListener
import com.apollographql.apollo.subscription.SubscriptionManagerState

class NoOpSubscriptionManager : SubscriptionManager {
  val errorMessage = "No `SubscriptionTransport.Factory` found, please add one to your `ApolloClient` with `ApolloClient.Builder.subscriptionTransportFactory`"

  override fun <T> subscribe(subscription: Subscription<*, T, *>, callback: SubscriptionManager.Callback<T>) {
    throw IllegalStateException(errorMessage)
  }

  override fun unsubscribe(subscription: Subscription<*, *, *>) {
    throw IllegalStateException(errorMessage)
  }

  override fun start() {
    throw IllegalStateException(errorMessage)
  }

  override fun stop() {
    throw IllegalStateException(errorMessage)
  }

  override fun reconnect() {
  }

  override fun getState(): SubscriptionManagerState {
    return SubscriptionManagerState.DISCONNECTED
  }

  override fun addOnStateChangeListener(onStateChangeListener: OnSubscriptionManagerStateChangeListener) {
    throw IllegalStateException(errorMessage)
  }

  override fun removeOnStateChangeListener(onStateChangeListener: OnSubscriptionManagerStateChangeListener) {
    throw IllegalStateException(errorMessage)
  }
}